library/core/src/main/java/com/google/android/exoplayer2/extractor/mp4/FragmentedMp4Extractor.java

```diff
@@ -1265,10 +1265,17 @@ private boolean readSample(ExtractorInput input) throws IOException, Interrupted
         sampleSize -= Atom.HEADER_SIZE;
         input.skipFully(Atom.HEADER_SIZE);
       }
-      sampleBytesWritten = currentTrackBundle.outputSampleEncryptionData();
-      sampleSize += sampleBytesWritten;
-      if (MimeTypes.AUDIO_AC4.equals(currentTrackBundle.track.format.sampleMimeType)) {
-        Ac4Util.getAc4SampleHeader(sampleSize, scratch);
+
+      boolean isAc4HeaderRequired =
+          MimeTypes.AUDIO_AC4.equals(currentTrackBundle.track.format.sampleMimeType);
+
+      int encryptionDataBytesWritten = currentTrackBundle.outputSampleEncryptionData(
+          sampleSize, isAc4HeaderRequired ? Ac4Util.SAMPLE_HEADER_SIZE : 0);
+      sampleBytesWritten = encryptionDataBytesWritten;
+      sampleSize += encryptionDataBytesWritten;
+
+      if (isAc4HeaderRequired) {
+        Ac4Util.getAc4SampleHeader(sampleSize - encryptionDataBytesWritten, scratch);
```

Comment:
Compared with your example code, the first parameter should be "sampleSize - encryptionDataBytesWritten", which is size of AC-4 RAW frame.